### PR TITLE
Range formatter

### PR DESCRIPTION
**Pull request by brevzin**
More complete implementation of [P2286](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2286r8.html) and [P2585](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2585r0.html) (renaming a couple things). For now this subsumes #2981.

This is still just the `n` specifier (no `s`, `?s`, or `m` yet) but at least it's the full structure now.

### CODE REVIEW

**Comment by brevzin** [comment]
I'll be honest I don't understand the gcc 4.8 error:

```
/home/runner/work/fmt/fmt/include/fmt/ranges.h:324:43: error: conversion from ‘const char*’ to non-scalar type ‘fmt::v9::basic_string_view<char>’ requested
```

What's wrong with [this constructor](https://github.com/fmtlib/fmt/blob/0db43cf7feb89f221f642c11b66bcc27043b2b5b/include/fmt/core.h#L450-L457)?

**Comment by mwinterb** [comment]
Looks like it's just an "ancient GCC with new features" bug. If you explicitly provide a default constructor for `formatter`, it seems like it'll compile. https://gcc.godbolt.org/z/Wcosaqa6P. `= default` won't work. Also, if you move the initialization of `separator_`, etc to that constructor, it'll work, too.

**Comment by mwinterb** [comment]
Also, to support the "esoteric" code unit types, I don't know if it'd be better to follow the `time_point` formatter model for specifying those three literals: https://github.com/fmtlib/fmt/blob/0db43cf7feb89f221f642c11b66bcc27043b2b5b/include/fmt/chrono.h#L2020-L2026
https://gcc.godbolt.org/z/q3KoTK9fK

**Comment by phprus** [comment]
``default_specs[]`` without defined array size breaks Intel compiler (and other with EDG frontend).
See fix for this bug: https://github.com/fmtlib/fmt/pull/2982

**Comment by mwinterb** [comment]
I was in a rush last night so I probably wasn't as clear as I should have been, but only `formatter() {}` would work.
However, you can also use brace initialization instead of assignment in the member initialization and that seems to be fine for 4.8: https://gcc.godbolt.org/z/E4boe4Gvr

**Review by vitaut — CHANGES_REQUESTED** [review]
Thanks for the PR. Please address build failures.

**Comment by brevzin** [comment]
> Thanks for the PR. Please address build failures.

Took me a while to figure this one out. Turns out this:

```
../include/fmt/core.h:2959:51: error: ‘constexpr decltype (ctx.begin()) fmt::v9::detail::parse_format_specs(ParseContext&) [with T = std::vector<int>; ParseContext = fmt::v9::detail::compile_parse_context<char, fmt::v9::detail::error_handler>; decltype (ctx.begin()) = const char*]’ called in a constant expression
 2959 |     return id >= 0 && id < num_args ? parse_funcs_[id](context_) : begin;
      |                                       ~~~~~~~~~~~~^
```

is a _slightly_ insufficient error diagnostic for "oh, you added a default constructor that you forgot to mark constexpr".

**Comment by phprus** [comment]
@brevzin what do you think about comment https://github.com/fmtlib/fmt/pull/2983#issuecomment-1181291453?

The ``FMT_STATICALLY_WIDEN`` macro is limiting the ``xchar`` support, including the standard ``char16_t`` and ``char32_t`` types.

**Comment by vitaut** [comment]
I think conversion is a better approach than having redundant literals for exotic code unit types.

**Comment by brevzin** [comment]
> I think conversion is a better approach than having redundant constants for exotic code unit types.

Converting what, where?

**Comment by vitaut** [comment]
I mean that instead of introducing a generic literal as suggested in https://github.com/fmtlib/fmt/pull/2983#issuecomment-1181291453 we could have a normal string literal and convert it into wide and other strings when needed.

**Comment by phprus** [comment]
The implementation of the ``set_separator`` and ``set_brackets`` requires that string literals must immediately be of the correct character type.

File: include/fmt/ostream.h [hunk 1]
```diff
@@ -129,14 +129,22 @@ template <typename T> struct streamed_view { const T& value; };
 
 // Formats an object of type T that has an overloaded ostream operator<<.
 template <typename Char>
-struct basic_ostream_formatter : formatter<basic_string_view<Char>, Char> {
+struct basic_ostream_formatter {
+private:
+  formatter<basic_string_view<Char>, Char> underlying_;
+
```
Comment (phprus) [inline include/fmt/ostream.h:135]: Why is this change needed?
Inheritance from formatter is allowed in the documentation (https://fmt.dev/latest/api.html):

> You can also reuse existing formatters via inheritance or composition, for example:
> ...
> ``template <> struct fmt::formatter<color>: formatter<string_view> {``
> ...
> Since parse is inherited from formatter<string_view> it will recognize all string format specifications, for example

Comment (brevzin) [reply include/fmt/ostream.h:135]: To avoid inheriting `set_debug_format()`

File: include/fmt/ostream.h [hunk 2]
```diff
@@ -129,14 +129,22 @@ template <typename T> struct streamed_view { const T& value; };
...
+  }
+
   template <typename T, typename OutputIt>
   auto format(const T& value, basic_format_context<OutputIt, Char>& ctx) const
       -> OutputIt {
     auto buffer = basic_memory_buffer<Char>();
     format_value(buffer, value, ctx.locale());
-    return formatter<basic_string_view<Char>, Char>::format(
-        {buffer.data(), buffer.size()}, ctx);
+    return underlying_.format({buffer.data(), buffer.size()}, ctx);
```
Comment (vitaut) [inline include/fmt/ostream.h:147]: This change looks unnecessary.

File: include/fmt/ranges.h [hunk 3]
```diff
@@ -391,56 +420,74 @@ template <typename R>
...
+        std::is_same<T, remove_cvref_t<T>>,
+        disjunction<is_formattable<T, Char>,
+                    detail::has_fallback_formatter<T, Char>>>::value>> {
+ private:
+  detail::range_formatter_type<Char, T> underlying_;
   bool custom_specs_ = false;
-  bool no_brackets_ = false;
+  basic_string_view<Char> separator_ = FMT_STATICALLY_WIDEN(Char, ", ");
+  basic_string_view<Char> opening_bracket_ = FMT_STATICALLY_WIDEN(Char, "[");
+  basic_string_view<Char> closing_bracket_ = FMT_STATICALLY_WIDEN(Char, "]");
```
Comment (vitaut) [inline include/fmt/ranges.h:440]: Let's do what was suggested in https://github.com/fmtlib/fmt/pull/2983#issuecomment-1181291453 and remove `FMT_STATICALLY_WIDEN`.

File: include/fmt/ranges.h [hunk 4]
```diff
@@ -391,56 +420,74 @@ template <typename R>
...
+
+  template <class U>
+  FMT_CONSTEXPR static void maybe_set_debug_format(U&, long) {}
+
+  FMT_CONSTEXPR void maybe_set_debug_format() {
+    maybe_set_debug_format(underlying_, 0);
+  }
+
+ public:
+  FMT_CONSTEXPR range_formatter() { }
```
Comment (vitaut) [inline include/fmt/ranges.h:456]: Please apply clang-format.

File: include/fmt/ranges.h [hunk 5]
```diff
@@ -391,56 +420,74 @@ template <typename R>
...
+  basic_string_view<Char> closing_bracket_ = FMT_STATICALLY_WIDEN(Char, "]");
+
+  template <class U>
+  FMT_CONSTEXPR static auto maybe_set_debug_format(U& u, int)
+      -> decltype(u.set_debug_format()) {
+    u.set_debug_format();
+  }
+
+  template <class U>
+  FMT_CONSTEXPR static void maybe_set_debug_format(U&, long) {}
```
Comment (vitaut) [inline include/fmt/ranges.h:449]: sneaky, we normally use ... for fallbacks like this.

**Review by vitaut — CHANGES_REQUESTED** [review]

File: include/fmt/ostream.h (hunk 2 above)
Comment (vitaut) [reply include/fmt/ostream.h:147]: Oh, I missed your earlier comment. Since we support all string specs here I don't see any reason to disable set_debug_format, please revert.

Comment (brevzin) [reply include/fmt/ostream.h:147]: The reason I did it was... let's say we have this:
```cpp
struct Int {
    int i;
    friend std:ostream& operator<<(std::ostream& os, Int i) { return os << i.i; }
};

std::vector<Int> ints = {Int{1}, Int{2}, Int{3}};
fmt::print("ints={}\n", ints);
```

If you keep the debug format, this'll print as `["1", "2", "3"]` (because the library thinks it's formatting string_views).

File: include/fmt/ranges.h [hunk 6]
```diff
@@ -450,75 +497,104 @@ struct formatter<
...
+template <typename T> struct range_format_kind_ {
+  static constexpr auto value = std::is_same<range_reference_type<T>, T>::value
+                                    ? range_format::disabled
+                                : is_map<T>::value ? range_format::map
+                                : is_set<T>::value ? range_format::set
+                                                   : range_format::sequence;
+};
+
+template <range_format K, typename R, typename Char, typename Enable = void>
+struct range_default_formatter;
```
Comment (vitaut) [inline include/fmt/ranges.h:532]: Why is it called `range_default_formatter` and not `default_range_formatter`? =)

File: include/fmt/ostream.h (hunk 2 above)
Comment (vitaut) [reply include/fmt/ostream.h:147]: Good point but can we disable `set_debug_format` more explicitly e.g. keep the inheritance but delete this function in ostream formatter?

File: include/fmt/ranges.h (hunk 6 above)
Comment (brevzin) [reply include/fmt/ranges.h:532]: I originally called it `default_range_formatter` (since in the paper I originally called it _`default-range-formatter`_) and then I got asked to change that one to _`range-default-formatter`_ so I did that... and now this just maps to what's in the paper.

File: include/fmt/ostream.h (hunk 2 above)
Comment (brevzin) [reply include/fmt/ostream.h:147]: When I did private inheritance, gcc 4.8 still picked it up anyway, but I didn't try `= delete;` which I agree is a more direct way of doing this. Will try and report back (or rather, if it works, I'll just change it. If it doesn't, then... I won't).

File: include/fmt/ranges.h (hunk 6 above)
Comment (vitaut) [reply include/fmt/ranges.h:532]: OK, let's keep it consistent with the paper.

File: include/fmt/ranges.h (hunk 3 above)
Comment (brevzin) [reply include/fmt/ranges.h]: Done, I think.

File: include/fmt/ostream.h (hunk 2 above)
Comment (brevzin) [reply include/fmt/ostream.h]: Looks like that worked.

File: include/fmt/ranges.h [hunk 7]
```diff
@@ -309,6 +309,16 @@ OutputIt write_range_entry(OutputIt out, const Arg& v) {
   return write<Char>(out, v);
 }
 
+template <typename CharT, CharT... C> struct string_literal {
+  static constexpr CharT value[] = {C...};
```
Comment (phprus) [inline include/fmt/ranges.h:313]: Please add array size for EDG based compilers. Details in PR https://github.com/fmtlib/fmt/pull/2982

```c++
static constexpr CharT value[sizeof...(C)] = {C...};
```

File: include/fmt/ranges.h [hunk 8]
```diff
@@ -391,56 +419,76 @@ template <typename R>
...
+      detail::string_literal<Char, ']'>{};
+
+  template <class U>
+  FMT_CONSTEXPR static auto maybe_set_debug_format(U& u, int)
+      -> decltype(u.set_debug_format()) {
+    u.set_debug_format();
+  }
+
+  template <class U>
+  FMT_CONSTEXPR static void maybe_set_debug_format(U&, long) {}
```
Comment (vitaut) [inline include/fmt/ranges.h:450]: long -> ... per earlier comment

Comment (brevzin) [reply include/fmt/ranges.h]: Oh I didn't realize you actually wanted me to change it. Sure.

File: include/fmt/ranges.h (hunk 7 above)
Comment (brevzin) [reply include/fmt/ranges.h]: Done.

**Review by vitaut — APPROVED** [review]

**Comment by vitaut** [comment]
Thanks!

**Comment by roman-orekhov** [comment]
Is this intentionally undocumented in [format syntax](https://fmt.dev/latest/syntax.html) until other specifiers are implemented as well?
